bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c

```diff
@@ -0,0 +1,622 @@
+    return 3;
+  }
+
+  // isinf() is a compiler intrinsic.
+  if (isinf(value)) {
+    strcpy(dst, "inf");
+    return 3;
+  }
+
+  // log10() is a compiler intrinsic.
+  int exponent = (int)log10(value);
+  // Values < 1.0 must subtract 1 from exponent to handle zero wraparound.
+  if (value < 1.0) {
+    exponent--;
+  }
+
+  // pow() is a compiler intrinsic.
+  double normalized = value / pow(10, exponent);
+  // Special case for 0.1, 0.01, 0.001, etc giving a normalized value of 10.xyz.
+  // We use 9.999... because 10.0 converts to a value > 10 in ieee754 binary
+  // floats.
+  if (normalized > 9.99999999999999822364316059975) {
+    exponent++;
+    normalized = value / pow(10, exponent);
+  }
+
+  // Put all of the digits we'll use into an integer.
+  double digits_and_remainder = normalized * pow(10, max_sig_digits - 1);
+  uint64_t digits = (uint64_t)digits_and_remainder;
+  // Also round up if necessary (note: 0.5 is exact in both binary and decimal).
+  if (digits_and_remainder - (double)digits >= 0.5) {
+    digits++;
+    // Special case: Adding one bumps us to next magnitude.
+    if (digits >= (uint64_t)pow(10, max_sig_digits)) {
+      exponent++;
+      digits /= 10;
+    }
+  }
+
+  // Extract the fractional digits.
+  for (int i = max_sig_digits; i > 1; i--) {
+    dst[i] = digits % 10 + '0';
+    digits /= 10;
+  }
+  // Extract the single-digit whole part.
+  dst[0] = (char)digits + '0';
+  dst[1] = '.';
+
+  // Strip off trailing zeroes, and also the '.' if there is no fractional part.
+  int e_offset = max_sig_digits;
+  for (int i = max_sig_digits; i > 0; i--) {
+    if (dst[i] != '0') {
+      if (dst[i] == '.') {
+        e_offset = i;
+      } else {
+        e_offset = i + 1;
+      }
+      break;
+    }
+  }
+  dst += e_offset;
+
+  // Add the exponent if it's not 0.
+  if (exponent != 0) {
+    *dst++ = 'e';
+    if (exponent >= 0) {
+      *dst++ = '+';
+    }
+    dst += int64_to_string(exponent, dst);
+  } else {
+    *dst = 0;
+  }
+
+  return (size_t)(dst - orig_dst);
+}
+
+/**
+ * Convert a positive double to a string, allowing up to max_sig_digits. See
+ * positive_double_to_string() for important information about how this
+ * algorithm differs from sprintf.
+ *
+ * This function will write a maximum of 22 characters (including the NUL) to
+ * dst.
+ *
+ * Returns the length of the string written to dst (not including the NUL).
+ */
+static size_t double_to_string(double value, char *dst, int max_sig_digits) {
+  if (value < 0) {
+    dst[0] = '-';
+    return positive_double_to_string(-value, dst + 1, max_sig_digits) + 1;
+  }
+  return positive_double_to_string(value, dst, max_sig_digits);
+}
+
+// ^^ End of copied code ^^
+
+// Max significant digits + special chars for sign, exponent, etc.
+#define MAX_STRING_SIZE_FLOAT 22
+
+static inline bool write_double(double value) {
+  char buff[MAX_STRING_SIZE_FLOAT];
+  double_to_string(value, buff, MAX_SIGNIFICANT_DIGITS);
+  return write_raw_string(buff);
```

Comment:
What is the performance impact of calling `write()` each time one of these methods is called? A very long time ago the Cocoa notifier did a similar thing and introducing a buffer to avoid `write()` each time drastically improved performance. It'd be good to confirm that this isn't introducing similar issues: https://github.com/bugsnag/bugsnag-cocoa/pull/160#issuecomment-324703817

Reply:
I've now replaced it with buffered_writer.c